### PR TITLE
Eip 4750/eof function sections

### PR DESCRIPTION
**Pull request by DemuirGos**
Fixes | Closes | Resolves #

> Anything marked as optional that you didn't need to fill in, please remove it from the PR description. Choose one of the keywords above to refer to the issue this PR solves, followed by the issue number (e.g Fixes # 666). If there is no issue, remove the line. Remove this note after reading.

## Changes:
- Implements Eip4750 Spec: 
  - Adds Opcodes (CALLF, RETF)
  - Adds Function like behavior to Bytecode
  - Adds Section checks to Eof checks
  - Deprecates JUMP/JUMPI, PC
  - Deprecates Jumpdest analysis

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

**Comments about testing , should you have some** (optional)

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm.Test/EOFChecks/EOF4750Tests.cs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+                    Bytecode = new ScenarioCase(
+                            Functions: new[] {
+                                new FunctionCase(
+                                    0, 0, 1024,
+                                    Prepare.EvmCode.Done
+                                )
+                            },
+                            Data: Bytes.FromHexString("deadbeef")
+                        ).Bytecode,
+                    Result = (StatusCode.Failure, "Invalid Code Section call"),
```
Comment (smartprogrammer93) [inline src/Nethermind/Nethermind.Evm.Test/EOFChecks/EOF4750Tests.cs:237]: `Empty Code Section`?

File: src/Nethermind/Nethermind.Evm/GasCostOf.cs [hunk 2]
```diff
@@ -19,6 +19,9 @@ public static class GasCostOf
         public const long RJump = 2;
         public const long RJumpi = 4;
         public const long RJumpv = 4;
+        public const long Callf = 5;
+        public const long Retf = 4;
+        public const long Jumpf = 4;
```
Comment (smartprogrammer93) [inline src/Nethermind/Nethermind.Evm/GasCostOf.cs:24]: yeet

**Review by smartprogrammer93 — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm.Test/BytecodeBuilderExtensionsTests.cs [hunk 3]
```diff
@@ -177,6 +179,11 @@ public static IEnumerable<TestCase> opcodes_with_1_arg()
 
             foreach (Instruction opcode in address_opcodes)
             {
+                if (_releaseSpec is Shanghai && opcode is Instruction.JUMPSUB)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm.Test/BytecodeBuilderExtensionsTests.cs:182]: hmm if EOF will be removed from shanghai we will have to change this code. What about IsEip4750Enabled?

File: src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs [hunk 4]
```diff
@@ -247,6 +250,52 @@ public bool TryParseEofHeader(ReadOnlySpan<byte> container, [NotNullWhen(true)]
...
+                if (_loggingEnabled)
+                {
+                    _logger.Trace($"EIP-3540 : CodeSection size must follow a CodeSection, CodeSection length was {codeSections.Length}");
+                }
+
+                header = null;
+                return false;
+            }
+
+            if (codeSections.Length != (typeSectionSize / 4))
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs:266]: what is for 4? avoid magic numbers

File: src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs [hunk 5]
```diff
@@ -247,6 +250,52 @@ public bool TryParseEofHeader(ReadOnlySpan<byte> container, [NotNullWhen(true)]
...
+                if (_loggingEnabled)
+                {
+                    _logger.Trace($"EIP-4750: first 2 bytes of type section must be 0s");
+                }
+
+                header = null;
+                return false;
+            }
+
+            if (codeSections.Length > 1024)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs:288]: 1024?

Comment (DemuirGos) [reply src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs]: 1024 is the Maximum number of allowed code sections in EOF bytecode

File: src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs (hunk 4 above)
Comment (DemuirGos) [reply src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs]: 4 is the minimum allowed size for typesections, which makes it the size a function typedata takes in the typesection (1 bytes for input count, 1 byte for output count, 2 bytes for max data stack height), I moved it to const field as well but I think MINIMUM_TYPESECTION_SIZE doesn't reflect what it does.

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 6]
```diff
@@ -3041,14 +3063,92 @@ void UpdateMemoryCost(in UInt256 position, in UInt256 length)
...
+                            if (!UpdateGas(GasCostOf.Callf, ref gasAvailable))
+                            {
+                                EndInstructionTraceError(EvmExceptionType.OutOfGas);
+                                return CallResult.OutOfGasException;
+                            }
+
+                            var index = (int)codeSection[programCounter..(programCounter + 2)].ReadEthUInt16();
+                            var inputCount = typeSection[index * 4];
+
+                            if (vmState.ReturnStackHead == 1024)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:3094]: Magic 1024

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 7]
```diff
@@ -3041,14 +3063,92 @@ void UpdateMemoryCost(in UInt256 position, in UInt256 length)
...
+                            }
+
+                            if (!UpdateGas(GasCostOf.Callf, ref gasAvailable))
+                            {
+                                EndInstructionTraceError(EvmExceptionType.OutOfGas);
+                                return CallResult.OutOfGasException;
+                            }
+
+                            var index = (int)codeSection[programCounter..(programCounter + 2)].ReadEthUInt16();
+                            var inputCount = typeSection[index * 4];
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:3092]: still a few magic numbers

File: src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs [hunk 8]
```diff
@@ -268,6 +271,52 @@ public bool TryParseEofHeader(ReadOnlySpan<byte> container, [NotNullWhen(true)]
...
+                if (_loggingEnabled)
+                {
+                    _logger.Trace($"EIP-4750: Code Sections count must match TypeSection count, CodeSection count was {codeSections.Length}, expected {typeSectionSize / MINIMUM_TYPESECTION_SIZE}");
+                }
+
+                header = null;
+                return false;
+            }
+
+            if (container[typeSectionStart] != 0 && container[typeSectionStart] != 0)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs:298]: Looks like bug. I guess it should be
`container[typeSectionStart] != 0 && container[typeSectionStart+1] != 0`

Comment (DemuirGos) [reply src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs:298]: yes I was fixing it it should be ``container[typeSectionStart] != 0 || container[typeSectionStart] != 0``

Comment (MarekM25) [reply src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs:298]: yes, || of course

**Pull request merged**